src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs

```diff
@@ -0,0 +1,82 @@
+            CreateLocalizableResourceString(nameof(ThreadStaticOnNonStaticFieldMessage)),
+            DiagnosticCategory.Usage,
+            RuleLevel.BuildWarning,
+            CreateLocalizableResourceString(nameof(ThreadStaticOnNonStaticFieldDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        // [ThreadStatic]
+        // private static object t_field = new object();
+        internal static readonly DiagnosticDescriptor ThreadStaticInitializedInlineRule = DiagnosticDescriptorHelper.Create(ThreadStaticInitializedInlineRuleId,
+            CreateLocalizableResourceString(nameof(ThreadStaticInitializedInlineTitle)),
+            CreateLocalizableResourceString(nameof(ThreadStaticInitializedInlineMessage)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(ThreadStaticInitializedInlineDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(ThreadStaticOnNonStaticFieldRule, ThreadStaticInitializedInlineRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(context =>
+            {
+                // Ensure ThreadStatic exists
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadStaticAttribute, out var threadStaticAttributeType))
+                {
+                    return;
+                }
+
+                // Warn on any [ThreadStatic] instance field.
+                context.RegisterSymbolAction(context =>
+                {
+                    var field = (IFieldSymbol)context.Symbol;
+                    if (field.HasAttribute(threadStaticAttributeType) && !field.IsStatic)
+                    {
+                        context.ReportDiagnostic(field.CreateDiagnostic(ThreadStaticOnNonStaticFieldRule));
+                    }
+                }, SymbolKind.Field);
+
+                // Warn on any [ThreadStatic] field inline initialization.
+                context.RegisterOperationAction(context =>
+                {
+                    var fieldInit = (IFieldInitializerOperation)context.Operation;
+                    foreach (IFieldSymbol field in fieldInit.InitializedFields)
+                    {
+                        if (field.IsStatic && field.HasAttribute(threadStaticAttributeType))
+                        {
+                            context.ReportDiagnostic(fieldInit.CreateDiagnostic(ThreadStaticInitializedInlineRule));
+                        }
+                    }
+
+                }, OperationKind.FieldInitializer);
```

Comment:
The loop here will probably result to unnecessary duplicate diagnostics. Consider checking one field only. I *think* all fields will have the same value for `IsStatic` and `.HasAttribute()`

Reply:
I changed it to exit the loop after the first diagnostic.